misc/python/materialize/data_ingest/executor.py

```diff
@@ -322,3 +334,133 @@ def run(self, transaction: Transaction) -> None:
                     else:
                         raise ValueError(f"Unexpected operation {row.operation}")
         self.pg_conn.commit()
+
+
+class KafkaRoundtripExecutor(Executor):
+    table: str
+    table_original: str
+    topic: str
+    known_keys: set[tuple[str]]
+
+    def __init__(
+        self,
+        num: int,
+        ports: dict[str, int],
+        fields: list[Field],
+    ):
+        super().__init__(ports)
+        self.table_original = f"table_rt_source{num}"
+        self.table = f"table_rt{num}"
+        self.topic = f"data-ingest-rt-{num}"
+        self.known_keys = set()
+
+        values = [
+            f"{field.name} {str(field.data_type.name(Backend.POSTGRES)).lower()}"
+            for field in fields
+        ]
+        keys = [field.name for field in fields if field.is_key]
+
+        self.mz_conn.autocommit = True
+        with self.mz_conn.cursor() as cur:
+            self.execute(cur, f"DROP TABLE IF EXISTS {self.table_original}")
+            self.execute(
+                cur,
+                f"""CREATE TABLE {self.table_original} (
+                        {", ".join(values)},
+                        PRIMARY KEY ({", ".join(keys)}));""",
+            )
+            self.execute(
+                cur,
+                f"""CREATE SINK sink{num} FROM {self.table_original}
+                    INTO KAFKA CONNECTION kafka_conn (TOPIC '{self.topic}')
+                    KEY ({", ".join(keys)})
+                    FORMAT AVRO
+                    USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                    ENVELOPE DEBEZIUM""",
+            )
+            self.execute(
+                cur,
+                f"""CREATE SOURCE {self.table}
+                    FROM KAFKA CONNECTION kafka_conn (TOPIC '{self.topic}')
+                    FORMAT AVRO
+                    USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                    ENVELOPE DEBEZIUM""",
+            )
+        self.mz_conn.autocommit = False
+
+    def run(self, transaction: Transaction) -> None:
+        with self.mz_conn.cursor() as cur:
+            for row_list in transaction.row_lists:
+                for row in row_list.rows:
+                    key_values = tuple(
+                        value
+                        for field, value in zip(row.fields, row.values)
+                        if field.is_key
+                    )
+                    if row.operation == Operation.INSERT:
```

Comment:
Optional: consider extracting methods for the different operations to reduce length and nesting. 🤓